flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/optimize/program/FlinkChangelogModeInferenceProgram.scala

```diff
@@ -0,0 +1,626 @@
+        // WindowAggregate, WindowTableAggregate, Deduplicate, TemporalSort, CEP, OverAggregate
+        // and WindowJoin require nothing about UpdateKind.
+        val children = visitChildren(rel, UpdateKindTrait.NONE)
+        createNewNode(rel, children, requiredTrait)
+
+      case rank: StreamExecRank =>
+        val rankStrategies = RankProcessStrategy.analyzeRankProcessStrategies(
+          rank, rank.partitionKey, rank.orderKey)
+        visitRankStrategies(rankStrategies, requiredTrait, rankStrategy => rank.copy(rankStrategy))
+
+      case sortLimit: StreamExecSortLimit =>
+        val rankStrategies = RankProcessStrategy.analyzeRankProcessStrategies(
+          sortLimit, ImmutableBitSet.of(), sortLimit.getCollation)
+        visitRankStrategies(
+          rankStrategies,
+          requiredTrait,
+          rankStrategy => sortLimit.copy(rankStrategy))
+
+      case sort: StreamExecSort =>
+        val requiredChildTrait = beforeAfterOrNone(getModifyKindSet(sort.getInput))
+        val children = visitChildren(sort, requiredChildTrait)
+        createNewNode(sort, children, requiredTrait)
+
+      case join: StreamExecJoin =>
+        val requiredUpdateBeforeByParent = requiredTrait.updateKind == UpdateKind.BEFORE_AND_AFTER
+        val children = join.getInputs.zipWithIndex.map {
+          case (child, childOrdinal) =>
+            val physicalChild = child.asInstanceOf[StreamPhysicalRel]
+            val needUpdateBefore = !join.inputUniqueKeyContainsJoinKey(childOrdinal)
+            val inputModifyKindSet = getModifyKindSet(physicalChild)
+            val childRequiredTrait = if (needUpdateBefore || requiredUpdateBeforeByParent) {
+              beforeAfterOrNone(inputModifyKindSet)
+            } else {
+              onlyAfterOrNone(inputModifyKindSet)
+            }
+            this.visit(physicalChild, childRequiredTrait)
+        }
+        if (children.exists(_.isEmpty)) {
+          None
+        } else {
+          createNewNode(join, Some(children.flatten.toList), requiredTrait)
+        }
+
+      case temporalJoin: StreamExecTemporalJoin =>
+        // forward required mode to left input
+        val left = temporalJoin.getLeft.asInstanceOf[StreamPhysicalRel]
+        val right = temporalJoin.getRight.asInstanceOf[StreamPhysicalRel]
+        val newLeftOption = this.visit(left, requiredTrait)
+        // currently temporal join only support insert-only source as the right side
+        // so it requires nothing about UpdateKind
+        val newRightOption = this.visit(right, UpdateKindTrait.NONE)
+        (newLeftOption, newRightOption) match {
+          case (Some(newLeft), Some(newRight)) =>
+            val leftTrait = newLeft.getTraitSet.getTrait(UpdateKindTraitDef.INSTANCE)
+            createNewNode(temporalJoin, Some(List(newLeft, newRight)), leftTrait)
+          case _ =>
+            None
+        }
+
+      case _: StreamExecCalc | _: StreamExecPythonCalc | _: StreamExecCorrelate |
```

Comment:
IIRC we've decided to not let `filter` passing the update kind. To solve the bad case like a single 'x < 10' condition which won't satisfy the `only_update_before` case.

Reply:
Sorry, I forgot to implement this conclusion.